cpp/include/raft/neighbors/ivf_pq_types.hpp

```diff
@@ -139,6 +151,90 @@ constexpr static uint32_t kIndexGroupSize = 32;
 /** Stride of the interleaved group for vectorized loads. */
 constexpr static uint32_t kIndexGroupVecLen = 16;
 
+/**
+ * Default value filled in the `indices()` array.
+ * One may encounter it trying to access a record within a cluster that is outside of the
+ * `list_sizes()` bound (due to the record alignment `kIndexGroupSize`).
+ */
+template <typename IdxT>
+constexpr static IdxT kInvalidRecord = std::numeric_limits<IdxT>::max() - 1;
+
+/**
+ * Default value returned by `search` when the `n_probes` is too small and top-k is too large.
+ * One may encounter it if the combined size of probed clusters is smaller than the requested
+ * number of results per query.
+ */
+template <typename IdxT>
+constexpr static IdxT kOutOfBoundsRecord = std::numeric_limits<IdxT>::max();
+
+/** PQ-encoded data stored in the interleaved format:
+ *
+ *    [ ceildiv(list_size, kIndexGroupSize)
+ *    , ceildiv(pq_dim, (kIndexGroupVecLen * 8u) / pq_bits)
+ *    , kIndexGroupSize
+ *    , kIndexGroupVecLen
+ *    ].
+ */
+template <typename SizeT = uint32_t>
+using list_extents =
+  extents<SizeT, dynamic_extent, dynamic_extent, kIndexGroupSize, kIndexGroupVecLen>;
+
+/** Determine the extents of an array enough to hold a given amount of data. */
+template <typename SizeT = uint32_t>
+constexpr auto make_list_extents(SizeT n_rows, uint32_t pq_bits, uint32_t pq_dim)
+  -> list_extents<SizeT>
+{
+  // how many elems of pq_dim fit into one kIndexGroupVecLen-byte chunk
+  auto pq_chunk = (kIndexGroupVecLen * 8u) / pq_bits;
+  return make_extents<SizeT>(div_rounding_up_safe<SizeT>(n_rows, kIndexGroupSize),
+                             div_rounding_up_safe<SizeT>(pq_dim, pq_chunk),
+                             kIndexGroupSize,
+                             kIndexGroupVecLen);
+}
+
+/** The data for a single list (cluster). */
+template <typename IdxT, typename SizeT = uint32_t>
+struct list_data {
+  /** PQ-encoded interleaved data. */
+  device_mdarray<uint8_t, list_extents<SizeT>, row_major> data;
+  /** Source indices. */
+  device_mdarray<IdxT, extent_1d<SizeT>, row_major> indices;
+  /** The actual size of the content. */
+  std::atomic<SizeT> size;
+
+  /** Allocate a new list capable of holding at least `n_rows` data records and indices. */
+  list_data(raft::device_resources const& res,
```

Comment:
Aside from the batch insert perf in FAISS, one of the reasons we need this is because we need to support conversion from and to the FAISS cpu and gpu layouts, which is not performant or trivial with a single block of memory. 

The faiss developers have created a class (the CodePacker) to allow us to encapsulate our layout pattern. I think instead of manually implementing the code packer logic for the raft indices in faiss, we should expose a couple little utility functions on our public API (maybe in a raft::neighbors::ivf_pq::helpers namespace like we did with kmeans) that would allow us to specify an index, a code size, a block size, some number of vectors, a chunk of memory with be flat codes, and provide a pointer to an output chunk of memory to write the packed "raft"-encoded vectors to. 

We could also provide another function that would accept the same input arguments but "unpack" a "raft"-encoded block back into its (non interleaved) flat form.

The code layout here is really a deep internal detail of our implementations and I'd prefer that we encapsulate that logic and keep it internal so that we can 

1) keep it close to the indexing code itself, and 

2) make future changes to it as needed  without having to manually change every library that's using raft downstream.

Reply:
List data is a part of the index declaration, so we have to keep it public. Since we're also going to use the same pattern in ivf-flat, I separated it into `ivf_list_types.hpp` and `ivf_list.hpp` and parameterized it with a `SpecT` template, which defines the implementation-specific data layout. I hope this make it more convenient to integrate with the CodePacker later!